Introducing Yaml configuration file

dipak-pawar opened the pull request:
#### Short description of what this resolves:
Added support for configuration file at parent location of multi-module project.

#### Changes proposed in this pull request:

- Introducing `smart-testing.yml`
```yml
mode: ordering
strategies: new, changed, affected
applyTo: surefire
debug: true
disable: false
report:
    enable: true
scm:
    range:
      head: HEAD
      tail: HEAD~3
```
- Introducing `Configuration.loadPrecalculated()` method to create configuration dumped at `.smart-testing/temporary/smart-testing.yml`
- Modified test-bed to use config file
 - Introducing fluent api to create config file using configuration object.
 ```java
 final Configuration configuration = Configuration.builder()
                .mode(SELECTING)
                .strategies(AFFECTED.getName())
                .scm()
                    .range()
                        .head("HEAD").tail("HEAD~")
                        .build()
                    .build()
                .build();
        project.configureSmartTesting()
                    .withConfiguration(configuration)
                .createConfigFile()
            .enable();
```


Fixes #113

Note: Documentation needs to update.

bartoszmajsak commented:
Thanks @dipak-pawar, I will go over it tomorrow morning (so in 12h from now ;)).

As you will be working on updating the documentation, which ideally should come as part of this PR, don't forget to cross-check all the constants which we are referring from there: https://github.com/arquillian/smart-testing/pull/183/files#diff-a1f054f1bd53a1e8081bd95bb2ce9447R5

Also we could think about using sample yaml config file for two purposes

  * for testing if we can unmarshall properly
  * the same file can be used as a living documentation - we can use tags to import it directly to the documentation, so when we introduce changes in the code and thus the file, we will have the documentation up-to-date. See how we leverage it for `Test Bed` for example. Including [`GitClone`](https://github.com/arquillian/smart-testing/blame/39c3e2cdf5e378953708488e7bbd346983031826/docs/developing.adoc#L20) directly from the [source](https://github.com/arquillian/smart-testing/blob/57e29f0d7d967ac07749a4d918d2e5ec564fbedb/functional-tests/git-rules/src/main/java/org/arquillian/smart/testing/rules/git/GitClone.java#L13-L15)

You can also use [footnotes](https://github.com/arquillian/smart-testing/blame/master/docs/developing.adoc#L83) to explain easily particular bits of the code.

bartoszmajsak commented:
```java
Configuration configuration = Configuration.builder()
               .mode(SELECTING)
               .strategies(AFFECTED.getName())
               .scm(Scm.builder().lastChanges("2").build())
           .build();
```

how about `scm()` returning builder instead of accepting it as a parameter? Would make it a bit more fluent.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,343 @@
+        return new Configuration.Builder();
+    }
+
+    public static class Builder {
```
If I understand the purpose of the builders correctly, they are used only for test purposes. If so, then I would move them to the `TestBed` implementation to keep the class clear and to not expose test logic from the normal class.

dipak-pawar replied:
@MatousJobanek If we move it to TestBed then we can't use it in unit test for [ConfigurationTest](https://github.com/arquillian/smart-testing/pull/183/files#diff-09137b69519fe19b2f7c61862edeba9b) & [ConfigurationUsingPropertyTest](https://github.com/arquillian/smart-testing/pull/183/files#diff-ebaf055f2f7537b003d0b8833f51d445). Also as we are keeping it here it's quite easy to reference things in future if we want to modify it.

MatousJobanek replied:
I'm still not a fan of this. If it is a test logic, then I would move it to the test-bed module or make it as hidden as possible (package protected). 
If you want to open the API to the world, then you should also verify that the builder works properly - create a test for the builder. 
If you moved it to the test-bed, then in Core you could use different approach of creating the object:
* reading a file
* using the set methods (that you have there anyway)
* configuring the object by providing `Properties` object - that would make more sense exposing it than the static builder API

dipak-pawar commented:
> how about scm() returning builder instead of accepting it as a parameter? Would make it a bit more fluent.

:+1: 

> **Passing strategies**
`.strategies(AFFECTED.getName())` why not AFFECTED anymore? Or is it just a mistake in the description?

No it's not mistake in the description. This is how I implemented it. I kn ow it's little weird.
We can add it.But the problem is we are using `Strategy` class from test-bed. & we have builder in `core` module. Either we have to Strategy to `core` or the other way around. But I don't want to use another way. as I won't use builder to test anything in core model if I move all builders to test-bed.

Also I am not sure it is good way to move `Strategy` class to `core`. So I implemented this way.

bartoszmajsak commented:
> Also I am not sure it is good way to move Strategy class to core. So I implemented this way.

Right, we keep strategies based on strings, not types, as we might want to open it up for new ones, when types would be pretty much blocking it.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,279 @@
+        if (files != null) {
+            if (files.length == 0) {
+                logger.info(
+                    "Config file `" + SMART_TESTING_YAML + "` OR `" + SMART_TESTING_YAML + "` is not found. Using system properties to load "
```
The usage of the same constant `SMART_TESTING_YAML` is correct?

MatousJobanek replied:
I guess that one of them should be `SMART_TESTING_YML`

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    public static Configuration load() {
+        final File parent = Paths.get("").toAbsolutePath().toFile();
```
I would be great to have option to define the root dir as a parameter of the method `load()`

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,279 @@
+            configuration.strategies = strategies.split("\\s*,\\s*");
+        }
+
+        configuration.disable = Boolean.valueOf(overWriteSystemProperty(yamlConfiguration, "disable", SMART_TESTING_DISABLE, "false"));
```
The default values `false` are not necessary for the `Boolean.valueOf`

dipak-pawar replied:
Here we are not using `false` as second parameter to `Boolean.valueof`, where as we are using it for `overwriteSystemPreoperty` method

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,279 @@
+            .get();
+    }
+
+    private static Configuration parseConfiguration(Map<String, Object> yamlConfiguration) {
```
This whole method is a little bit chaotic to my taste. I'm slightly afraid of adding additional props and also what it will look like when we support also configuration per strategy or when the size of parameters will grow. 
Have you considered my comment: https://github.com/arquillian/smart-testing/issues/113#issuecomment-328831162
Similar approach could be also useful for you - having one class "Configuration" and go through all variables stored in this class, based on the type and name read relative parameters from the file and store inside of the class instance. If the variable is something different than array/list/string/int/boolean and similar (in other words it is some additional object such as `Report`), then take this new class and do the same stuff also for this one... 
You know, some kind of generic engine that takes the yaml file and any Object and based on the structure read and fill corresponding values. 
Then, adding/removing parameters would be as easy as modifying the class structure.
Something similar I do for browser options in Drone: https://github.com/arquillian/arquillian-extension-drone/blob/master/drone-webdriver/src/main/java/org/jboss/arquillian/drone/webdriver/factory/CapabilitiesOptionsMapper.java
I think that you could find similar code also in Arquillian Core or other project...

MatousJobanek commented on surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/SmartTestingProviderTest.java:
```diff
@@ -101,6 +113,17 @@ public void test_when_invoke_is_called_with_whole_set_of_classes() throws Except
+    @AfterClass
+    public static void purgeConfiguration() {
+        final Path tempDirPath = Paths.get(Paths.get("").toAbsolutePath().toString(), SMART_TESTING_WORKING_DIRECTORY_NAME);
+        FileSystemOperations.deleteDirectory(tempDirPath, true);
```
wouldn't be better to use `TemporaryFolder` rule?

dipak-pawar replied:
We can't use it `TemporaryFolder` Rule because it has it's own format for naming `given_name_time_stamp`. We need to dump configuration to `${projectdir}/.smart-testing/temporary/smart-testing.yml` So `Configuration.loadPrecalculated(File projectDir)` method can load configuration from that predefined location.

MatousJobanek replied:
setting the path to the project root directory should solve this problem - see my comment: https://github.com/arquillian/smart-testing/pull/183#discussion_r140525783

MatousJobanek commented on mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java:
```diff
@@ -54,17 +59,38 @@ public void afterProjectsRead(MavenSession session) throws MavenExecutionExcepti
 
+    private void storeConfiguration(MavenSession session) {
+        final Path path = configuration.dump();
+        session.getAllProjects().forEach(mavenProject -> copyConfigurationFile(mavenProject.getModel(), path.toFile()));
```
We already go through all projects in `configureExtension(session, configuration);` - don't loop twice

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    public Path dump() {
+        final LocalStorageFileAction configFile = new LocalStorage(Paths.get("").toFile())
```
setting path to the root dir should be done using a parameter in `dump()`;

MatousJobanek replied:
The same can be applied also to the other methods

MatousJobanek replied:
mainly the `loadPrecalculated()` one - otherwise it won't work in the surefire forked mode

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,279 @@
+    static Configuration loadConfigurationFromFile(File configFile) {
+        try(FileReader fileReader = new FileReader(configFile)) {
+            final Yaml yaml = new Yaml();
+            return yaml.loadAs(fileReader, Configuration.class);
```
now, I have noticed this line... Why don't you use this method also for loading the yaml file configured by the user? 
related to my comment: https://github.com/arquillian/smart-testing/pull/183/files#r140521368

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/hub/storage/local/FileSystemOperations.java:
```diff
@@ -62,7 +62,7 @@ static void deleteFile(Path path, boolean catchException) {
     }
 
-    static void deleteDirectory(Path directory, boolean catchException) {
+    public static void deleteDirectory(Path directory, boolean catchException) {
```
doesn't have to be public - see my comment: https://github.com/arquillian/smart-testing/pull/183/files#r140523705

MatousJobanek commented on core/src/test/java/org/arquillian/smart/testing/impl/TestStrategyApplierUsingPropertyTest.java:
```diff
@@ -135,4 +140,9 @@ private TestExecutionPlannerLoader prepareLoader(final Set<Class<?>> testsToRun,
         return new LinkedHashSet<>(Arrays.<Class<?>>asList(testClasses));
     }
+
+    private String getBaseDir(Class<?> aClass) {
```
do we really need this?

dipak-pawar replied:
yes we need this, otherwise we have to hard code this value to "core" to work it from `IDEA` & terminal.

lordofthejars commented:
Are we far from merging this?

dipak-pawar commented:
> Are we far from merging this?

We are not too far from merging. Needs final review from @MatousJobanek

dipak-pawar commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,267 @@
+    private static Configuration parseConfiguration(Map<String, Object> yamlConfiguration) {
+        final Configuration configuration = ObjectMapper.mapToObject(Configuration.class, yamlConfiguration);
+        ObjectMerger<Configuration> configurationMerger =
+            new ObjectMerger<>(Configuration.fromSystemProperties(), configuration, Configuration.withDefaultValues());
```
yes that's what we are doing in `ObjectMerger.merge()`

dipak-pawar commented:
> One general question, is configuration information written as report?

no it's not. but as this story is too big, I think it's good to create small task for it, after we merge this.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/ObjectMapper.java:
```diff
@@ -0,0 +1,133 @@
+
+public class ObjectMapper {
+
+    public static <T> T mapToObject(Class<T> aClass, Map<String, Object> map) {
```
Do we really need to have it public? 
This is used only by the `Configuration` class, so I would move it to the `configuration` package and make it package protected. If we need it also in some other cases, then we can open it to the world.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/ObjectMerger.java:
```diff
@@ -0,0 +1,127 @@
+    private T second;
+    private T third;
+
+    public ObjectMerger(T first, T second, T third) {
```
Same as for `ObjectMapper` - do not expose our implementation details if not necessary

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,267 @@
+        Map<String, Object> yamlConfiguration = new LinkedHashMap<>();
+
+        if (files != null) {
+            if (files.length == 0) {
```
these two `if`s should be merged, to handle (log) also the case when `files` is `null`

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/ObjectMapper.java:
```diff
@@ -0,0 +1,133 @@
+            return handleEnum(method, mappedValue);
+        } else if (parameterType.isAssignableFrom(List.class)) {
+            return handleList(method, mappedValue);
+        } else if (!parameterType.isAssignableFrom(Map.class) && Map.class.isAssignableFrom(mappedValue.getClass())) {
```
This if should check other configuration objects such as `Report` and `Scm`. Personally, I would create an interface (eg. `ConfigurationSection`) that all configuration classes would implement it... Then you could check if the parameter type is assignable from the interface.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,267 @@
+        configuration.applyTo = System.getProperty(SMART_TESTING_APPLY_TO);
+
+        configuration.report = Report.fromSystemProperties();
+        configuration.scm = Scm.fromSystemProperties();
```
To be honest, I still don't like this tightly coupled way of merging the configurations. I just cannot imagine the solution for separated configurations for each strategy. These strategy related configs should be part of the strategy modules and this doesn't allow it. 
I would go for the solution I was proposing you some time ago - having kind of list of objects that contains all necessary information: param name, system property, default value. Then, based on the type (implementing an interface), you could call the method that returns the list of the mentioned objects. This would simplify the logic of getting the system properties (making it generic for all parameters), you could get rid of the `ObjectMapper` class and also simplify this [if](https://github.com/arquillian/smart-testing/pull/183/files#diff-550791242a7d1b9809c8c5ed3e9ac5ecR60). At the same time you could reuse the [convertToType](https://github.com/arquillian/smart-testing/pull/183/files#diff-550791242a7d1b9809c8c5ed3e9ac5ecR94) method for getting the system properties. 
We can have a bjn tomorrow and I can show you what I'm talking about. Some pair programming would make sense as well - I have already some bits implemented.

MatousJobanek replied:
here is the implementation I'm talking about. https://github.com/MatousJobanek/smart-testing/commit/17f94b647ff0feb8b211d8fee9e8d99972d83e62 You can take a look so you know what I have in my mind. It needs some refactoring, removing old implementation classes and methods (such as `fromSystemProperties`) and changes in test classes, but it should work. 
I believe that it will make it simpler and easier to use. 
In addition, I haven't found (in your code) the solution for reading the list of parameters & it's default methods - for Che integration we were talking about. Having the list of configuration items for every configuration section should make it easier to implement.
Let's discuss it in the morning

dipak-pawar replied:
you can find solution for reading the list of parameters in [SchemaGenerator](https://github.com/arquillian/smart-testing/pull/183/files#diff-01b186fdde2ad7c30adc1cfbeb7b08f9)

MatousJobanek replied:
Yeah, I missed that class. Probably because it is package protected and as I need to call it from Che logic I didn't pay much attention to it.
Is there some way of getting default values?

dipak-pawar replied:
You mean default values for each property. for this we need to keep some map with `key` as `property(class's instance field)` & `value` as `default value` for that field.

MatousJobanek replied:
As for the `SchemaGenerator` - maybe it would be good to remove it from this PR and finish it in a separated PR - we can do it together.
This PR is huge enough, so let's finish the main logic first.
I'll create an issue for it

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Range.java:
```diff
@@ -0,0 +1,24 @@
+package org.arquillian.smart.testing.configuration;
+
+public class Range {
```
Why do we have a separated class for the `Range`? What is the difference between having `Scm` + `Range` and only `ScmRange`?

dipak-pawar replied:
The main idea to keep both classes separated to keep same structure in system properties  (`scm.range.head` & `scm.range.tail`) & in yaml
```yaml
scm:
    range:
      head: HEAD
      tail: HEAD~3
```

Otherwise we have to give option in config file as 
```yaml
scmRange:
      head: HEAD
      tail: HEAD~3
```

Also for this second option, we need to think about `lastChanges` option, as it's defined as `scm.lastChanges` as system prioperty for now. So either we need to change it to `scmRange.lastChanges` if we decided to go with second option

MatousJobanek commented on core/src/test/resources/dumped-smart-testing.yml:
```diff
@@ -0,0 +1,9 @@
+!!org.arquillian.smart.testing.configuration.Configuration
```
the directory ` core/src/test/resources` is getting to be chaotic - move the config files to separated subdirectory.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/api/SmartTesting.java:
```diff
@@ -23,7 +24,7 @@
      */
     public static ConfiguredSmartTesting with(TestVerifier testVerifier) {
-        return with(testVerifier, Configuration.load());
+        return with(testVerifier, Configuration.loadPrecalculated(Paths.get("").toFile()));
```
As a default behavior, I would load the file from main directory - the original one. `Configuration.load(Paths.get("").toFile())`
The precalculated is Maven(surefire) specific.

dipak-pawar replied:
No this `with` method is not used anywhere. We are using with `(TestVerifier testVerifier, Configuration configuration)`. You can look [here](https://github.com/arquillian/smart-testing/pull/183/files/99430c1dc61dcd34884d69229d0eaf08c30e1210#diff-e92614f0bf12fbca05272c1456ecba67R67). I am not sure about `che` for this. Is it used or not?

Also one more thing, as with default pointing to `rootDir` other submodule won't have this configuration file in it's respective directory, So it won't work

MatousJobanek replied:
It had been used somewhere in ST, but not now obviously. 
BTW, the API is not intended to be used for one parent module expecting that it will invoke all submodules.
Nevermind, leave it as it is.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,234 @@
+        return ObjectMapper.mapToObject(Configuration.class, yamlConfiguration);
+    }
+
+    private static boolean containsAnyStrategy(String strategies) {
```
not used

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/ObjectMapper.java:
```diff
@@ -0,0 +1,177 @@
+        return instance;
+    }
+
+    private static <T> void invokeMethodWithMappedValue(List<ConfigurationItem> configItems, Method method, T instance,
```
maybe break this method not to be so long - something like separating: https://gist.github.com/MatousJobanek/9e79ba640b3b98361c2afca283aa867a

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/ObjectMapper.java:
```diff
@@ -0,0 +1,177 @@
+            }
+            return mapToObject((Class<ConfigurationSection>) parameterType, (Map<String, Object>) mappedValue);
+        } else if (parameterType.isAssignableFrom(mappedValue.getClass())) {
+            return mappedValue;
```
now looking at it - moving the 
```java
} else if (parameterType.isAssignableFrom(mappedValue.getClass())) {
            return mappedValue;
```
on the line 93 would eliminate the nested if on the lines 94-96

dipak-pawar replied:
no they both covers different scenarios. If you run `ObjectMapperTest` you will get clear idea why I did this way

MatousJobanek replied:
I changed it to:
```java
        if (parameterType.isArray()) {
            return handleArray(parameterType.getComponentType(), mappedValue);
        } else if (parameterType.isEnum()) {
            return handleEnum(method, mappedValue);
        } else if (parameterType.isAssignableFrom(List.class)) {
            return handleList(method, mappedValue);
        } else if (parameterType.isAssignableFrom(mappedValue.getClass())) {
            return mappedValue;
        } else if (ConfigurationSection.class.isAssignableFrom(parameterType)) {
            return mapToObject((Class<ConfigurationSection>) parameterType, (Map<String, Object>) mappedValue);
        } else {
            return convertToType(parameterType, mappedValue.toString());
        }
```
and the tests are green

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Scm.java:
```diff
@@ -0,0 +1,43 @@
+
+    private Range range;
+
+    public void setLastChanges(String lastChanges) {
```
not used

dipak-pawar replied:
It's used by `ObjectMapper`

MatousJobanek replied:
now I got your point... interesting idea - I didn't think about it in this way. 👍

MatousJobanek replied:
add the annotation `@SuppressWarnings("unused")`

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java:
```diff
@@ -0,0 +1,234 @@
+    private static final Logger logger = Logger.getLogger();
+
+    public static final String DEFAULT_MODE = "selecting";
+    public static final String DEFAULT_STRATEGIES = "";
```
both variables `DEFAULT_MODE` and `DEFAULT_STRATEGIES` are not used

dipak-pawar replied:
`DEFAULT_MODE` is used in documentation. `DEFAULT_STRATEGIES` are used anywhere

MatousJobanek replied:
then use the `DEFAULT_MODE` also here https://github.com/arquillian/smart-testing/pull/183/files#diff-348861c21efd28f5cc99117b2c67302aR110

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/ConfigurationItem.java:
```diff
@@ -0,0 +1,40 @@
+    private String systemProperty;
+    private Object defaultValue;
+
+    public ConfigurationItem(String paramName, String systemProperty, Object defaultValue) {
```
the constructors would deserve Javadoc - mainly describing the `paramName` - that it needs to match the set methods

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/configuration/Report.java:
```diff
@@ -0,0 +1,48 @@
+
+    private boolean enable;
+    private String name = REPORT_FILE_NAME;
+    private String dir = TARGET;
```
if the defaults are in `registerConfigurationItems` then they don't have to be here... just a detail...

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/SchemaGenerator.java:
```diff
@@ -0,0 +1,89 @@
+        map.put(TYPE, "object");
+        map.put("javatype", aClass.getName());
+
+        final List<Field> list = Arrays.stream(aClass.getDeclaredFields())
```
here you get the information about used params from the declared fields, but when we load the config, we use setter methods. Eg the `setLastChanges` won't be covered here.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/SchemaGenerator.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+class SchemaGenerator {
```
if the SchemaGenerator is related to configuration, then move it there - into the same package

MatousJobanek commented on core/src/test/resources/configuration/smart-testing-1.yml:
```diff
@@ -0,0 +1,12 @@
+mode: ordering
```
I would welcome some better suffix then just `1`

MatousJobanek commented on docs/configuration.adoc:
```diff
@@ -57,15 +57,15 @@ instead of *new* ones.
 between commits range and local changes as well.
 
-TIP: `const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="COMMIT"]`, `const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="PREVIOUS_COMMIT"]` and `const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="LAST_COMMITS"]` Java system properties are valid as well.
+TIP: `const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="SCM_RANGE_HEAD"]`, `const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="SCM_RANGE_TAIL"]` and `const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="DEFAULT_LAST_COMMITS"]` Java system properties are valid as well.
```
I guess that the `DEFAULT_LAST_COMMITS` is wrong - there should be the system property used...

MatousJobanek commented on docs/usage.adoc:
```diff
@@ -8,22 +8,22 @@ or `maven-failsafe-plugin`.
 
 [[orderingChanged]]
-`mvn clean test -Dconst:core/src/main/java/org/arquillian/smart/testing/Configuration.java[name="SMART_TESTING_MODE"]=ordering -Dconst:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="LAST_COMMITS"]=1 -Dconst:core/src/main/java/org/arquillian/smart/testing/Configuration.java[name="SMART_TESTING"]="new, changed"`  copyToClipboard:orderingChanged[]
+`mvn clean test -Dconst:core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java[name="SMART_TESTING_MODE"]=ordering -Dconst:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="DEFAULT_LAST_COMMITS"]=1 -Dconst:core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java[name="SMART_TESTING"]="new, changed"`  copyToClipboar ...
```
same here with the `DEFAULT_LAST_COMMITS`

MatousJobanek commented on docs/jenkins.adoc:
```diff
@@ -59,7 +59,7 @@ pipeline {
                 script {
                     def scmVars = checkout scm
-                    sh "mvn -Dconst:core/src/main/java/org/arquillian/smart/testing/Configuration.java[name="SMART_TESTING"]='new, affected' -Dconst:core/src/main/java/org/arquillian/smart/testing/Configuration.java[name="SMART_TESTING_MODE"]=ordering -Dconst:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="COMMIT"]=${scmVars.GIT_COMMIT} -Dconst:core/src/main/jav ...
+                    sh "mvn -Dconst:core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java[name="SMART_TESTING"]='new, affected' -Dconst:core/src/main/java/org/arquillian/smart/testing/configuration/Configuration.java[name="SMART_TESTING_MODE"]=ordering -Dconst:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="SCM_RANGE_HEAD"]=${scmVars.G ...
```
missing `T` in `${scmVars.GIT_PREVIOUS_COMMI}`

MatousJobanek commented on docs/refcard.adoc:
```diff
@@ -42,17 +42,17 @@ a|`surefire`, `failsafe`
-a|`const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="PREVIOUS_COMMIT"]`
+a|`const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="SCM_RANGE_TAIL"]`
 |Set first commit id for inspecting changes
 a|`const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="HEAD"]~const:core/src/main/java/org/arquillian/smart/testing/scm/ScmRunnerProperties.java[name="DEFAULT_LAST_COMMITS"]`
```
again `DEFAULT_LAST_COMMITS`

MatousJobanek commented on core/src/test/java/org/arquillian/smart/testing/impl/TestStrategyApplierUsingPropertyTest.java:
```diff
@@ -135,4 +140,9 @@ private TestExecutionPlannerLoader prepareLoader(final Set<Class<?>> testsToRun,
         return new LinkedHashSet<>(Arrays.<Class<?>>asList(testClasses));
     }
+
+    private String getBaseDir() {
```
you told me:

> yes we need this, otherwise we have to hard code this value to "core" to work it from IDEA & terminal.

I tried it with these options:
* `System.getProperty("user.dir")`
* `new File(".").getAbsolutePath()`
* removing the `in(getBaseDir())` completely

and in all cases the tests are green on both command line and IDE

dipak-pawar replied:
I know why it's passing because you added fix for this few days ago in #181

MatousJobanek commented on docs/configfile.adoc:
```diff
@@ -0,0 +1,52 @@
+== Configuration file
+
+You can use smart-testing.y(a)ml file which is present at `${project.dir}/smart-testing.yml`.
```
`smart-testing.y(a)ml` this doesn't look obvious in the doc. Specify both types or just write:

You can use smart-testing yaml file at `${project.dir}/smart-testing.yml` or with alternative suffix `${project.dir}/smart-testing.yaml`

MatousJobanek replied:
now I have noticed that it is mentioned here https://github.com/arquillian/smart-testing/pull/183/files#diff-1dfff5bec990830d634a922b323a0661R18 again

MatousJobanek commented on docs/configfile.adoc:
```diff
@@ -0,0 +1,52 @@
+
+NOTE: You can use either a .yml or .yaml extension for this file.
+
+==== strategies
```
instead of creating a new titli for every param I would use the same approach of explaining parameters as in cube config: http://arquillian.org/arquillian-cube/#_configuration

MatousJobanek replied:
you know, referencing from the config file example using numbers

MatousJobanek commented:
Good job @dipak-pawar 👍 
Thank you for your great work 🥇
